compiler/dyno/tools/chpldoc/chpldoc.cpp

```diff
@@ -792,190 +897,289 @@ struct RstResult {
+    if (isNoDoc(node)) return false;
 
-    os_ << ".. " << kind << ":: ";
+    if (!textOnly_) os_ << ".. " << kind << ":: ";
     RstSignatureVisitor ppv{os_};
     node->traverse(ppv);
-    os_ << "\n";
-
-    showComment(node, indentComment);
+    if (!textOnly_) os_ << "\n";
+    bool commentShown = showComment(node, indentComment);
+
+    // TODO: Fix all this because why are we checking for specific node types?
+    if (commentShown && !textOnly_ && (node->isEnum() ||
+                                       node->isClass() ||
+                                       node->isRecord() ||
+                                       node->isModule())) {
+      os_ << "\n";
+    }
 
     if (auto attrs = node->attributes()) {
-      if (attrs->isDeprecated()) {
-        indentStream(os_, 1 * indentPerDepth) << ".. warning::\n";
-        indentStream(os_, 2 * indentPerDepth) << attrs->deprecationMessage().c_str();
-        os_ << "\n\n";
+      if (attrs->isDeprecated() && !textOnly_) {
+        auto comment = previousComment(context_, node->id());
+        if (comment && !comment->str().empty() &&
+            comment->str().substr(0, 2) == "/*" &&
+            comment->str().find("deprecat") != std::string::npos ) {
+            // do nothing because deprecation was mentioned in doc comment
+        } else {
+          // write the deprecation warning and message
+          indentStream(os_, indentDepth_ * indentPerDepth) << ".. warning::\n\n";
+          indentStream(os_, (indentDepth_ + 1) * indentPerDepth);
+          if (attrs->deprecationMessage().isEmpty()) {
+            // write a generic message because there wasn't a specific one
+            os_ << getNodeName((AstNode*) node) << " is deprecated";
+          } else {
+            // use the specific deprecation message
+            os_ << strip(attrs->deprecationMessage().c_str());
+          }
+          os_ << "\n\n";
+        }
       }
     }
+    return commentShown;
   }
 
   void visitChildren(const AstNode* n) {
+    std::vector<RstResult*> subModules;
     for (auto child : n->children()) {
       if (auto &r = rstDoc(context_, child->id())) {
-        children_.push_back(r.get());
+        if (child->isModule()) {
+          if (!writeStdOut_) {
+            // lookup the module path
+            std::vector<UniqueString> modulePath = getModulePath(context_,
+                                                                 child->id());
+            modulePath.pop_back();
+            std::string parentPath = modulePath.back().str();
+            std::string outdir = outputDir_ + "/" + parentPath;
+            r->outputModule(outdir, child->toModule()->name().str(),
+                            indentPerDepth);
+
+          } else {
+            subModules.push_back(r.get());
+          }
+        } else {
+          children_.push_back(r.get());
+        }
       }
     }
+    // add the submodules to the end
```

Comment:
I'll update the comment but the short answer is because the docs expect all of a single module's fields/functions to be printed together (grouped by module). Without moving the modules to the end, we intermix the parent modules fields/functions with the child module's docs